src/node_contextify.cc

```diff
@@ -1299,10 +1335,97 @@ void ContextifyContext::CompileFunction(
+    const FunctionCallbackInfo<Value>& args) {
+  // Argument 1: source code
+  CHECK(args[0]->IsString());
+  Local<String> code = args[0].As<String>();
+
+  // Argument 2: filename
+  CHECK(args[1]->IsString());
+  Local<String> filename = args[1].As<String>();
+
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+  // TODO: Centralize this rather than matching the logic in cjs/loader.js and translators.js
+  Local<Symbol> id_symbol = (String::Concat(isolate,
+      String::NewFromUtf8(isolate, "cjs:").ToLocalChecked(), filename)).As<Symbol>();
+
+  // TODO: Abstract this into a separate function
+  // Set host_defined_options
+  Local<PrimitiveArray> host_defined_options =
+      PrimitiveArray::New(isolate, loader::HostDefinedOptions::kLength);
+  host_defined_options->Set(
+      isolate, loader::HostDefinedOptions::kID, id_symbol);
+
+  ScriptOrigin origin(isolate,
+                      filename,
+                      0,                 // line offset
+                      0,                 // column offset
+                      true,              // is cross origin
+                      -1,                // script id
+                      Local<Value>(),    // source map URL
+                      false,             // is opaque (?)
+                      false,             // is WASM
+                      false,             // is ES Module
+                      host_defined_options);
+
+  ScriptCompiler::CachedData* cached_data = nullptr;
+  ScriptCompiler::Source source = ScriptCompiler::Source(code, origin, cached_data);
+  ScriptCompiler::CompileOptions options;
+  if (source.GetCachedData() == nullptr) {
+    options = ScriptCompiler::kNoCompileOptions;
+  } else {
+    options = ScriptCompiler::kConsumeCodeCache;
+  }
+  // End TODO
+
+  std::vector<Local<String>> params;
+  params.push_back(String::NewFromUtf8(isolate, "exports").ToLocalChecked());
+  params.push_back(String::NewFromUtf8(isolate, "require").ToLocalChecked());
+  params.push_back(String::NewFromUtf8(isolate, "module").ToLocalChecked());
+  params.push_back(String::NewFromUtf8(isolate, "__filename").ToLocalChecked());
+  params.push_back(String::NewFromUtf8(isolate, "__dirname").ToLocalChecked());
+
+  TryCatchScope try_catch(env);
+
+  ContextifyContext::CompileFunctionAndCacheResult(
+    env,
+    context,
+    source,
+    params,
+    std::vector<Local<Object>>(),
+    options,
+    true,
+    id_symbol,
+    try_catch);
+
+  if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
+    Utf8Value message_value(env->isolate(), try_catch.Message()->Get());
+    auto message = message_value.ToStringView();
+
+    for (const auto& error_message : esm_syntax_error_messages) {
+      if (message.find(error_message) != std::string_view::npos) {
+        args.GetReturnValue().Set(true);
+        break;
+      }
+    }
     return;
```

Comment:
This return is wrong. It returns undefined not false because your tests check for falsy statements.